ir/terminator.go

```diff
@@ -109,7 +109,7 @@ func NewBr(target *Block) *TermBr {
 func (term *TermBr) Succs() []*Block {
 	// Cache successors if not present.
 	if term.Successors == nil {
-		term.Successors = []*Block{term.Target}
+		term.Successors = []*Block{term.Target.(*Block)}
```

Comment:
So, in this case, we can use `Successors []ir.BlockInterface`. At line 149 to 151:

```go
func NewCondBr(cond value.Value, targetTrue, targetFalse *Block) *TermCondBr {
	return &TermCondBr{Cond: cond, TargetTrue: targetTrue, TargetFalse: targetFalse}
}
```

A problem would be: although we help users can use `value.Value` as `TargetTrue/False`, but the function signature still denied a general type, so they have to modify the field and lose the type guarantee.

Reply:
> the function signature still denied a general type, so they have to modify the field and lose the type guarantee.

This is intended, as the underlying type of `Target`, `TargetTrue` and `TargetFalse` is always `*ir.Block`. The only reason we change the type to `value.Value` is to allow for the use-tracking API to take the pointer to these values (`*value.Value`).

It is not a perfect solution, but considering benefits and drawbacks, I think that the benefits outweigh the drawbacks, as using `value.Value` for operands and branch targets of instructions and terminators enable quite a powerful use-tracking API.

Having a constructor with the restricted types help ensure that users still provide a value `value.Value` type for these operands and branch targets.

Reply:
I rethink this part: a problem is `Successors` makes `Target` have to do type casting, which might panic at the runtime if users didn't notice this thing. Seems like it should also use `[]value.Value` type(following current PR context).

Reply:
I rethink this part: a problem is `Successors` makes `Target` have to do type casting, which might panic at the runtime if users didn't notice this thing. Seems like it should also use `[]value.Value` type(following current PR context).

Reply:
> I rethink this part: a problem is Successors makes Target have to do type casting, which might panic at the runtime if users didn't notice this thing. Seems like it should also use []value.Value type(following current PR context).

A panic at runtime is to prefer in this case as using anything other than `*ir.Block` for a branch target would result in an invalid IR. As noted above, we would then rely on semantic analysis of the IR module to report an error to the user of such invalid IR instances. And furthermore, the terminator constructors (e.g. `ir.NewBr`) already restrict the type of branch targets to `*ir.Block`, so as long as users use the constructor functions of `llir/llvm/ir`, then creating an invalid IR is made more difficult.

In essence, the idea is to have types be as specific (and restrictive) as possible where there are no other alternatives. That is the case for the parameters of e.g. `ir.NewBr` which take a `target *ir.Block` parameter. And that is also the case for `Successors` which is later used in the `Succs` method, the return type of which is `[]*ir.Block`.

I know it's not perfect, but it is a trade-off, so we can have a more general and powerful use-tracking API which operates on `*value.Value`.